File: 16-3sum-closest/3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        int sum = 0;
        int closest = Integer.MAX_VALUE;
        int ret = 0;

        Arrays.sort(nums);

        for(int i = 0; i < nums.length - 2; i++) {
            int left = i + 1;
            int right = nums.length - 1;

            while(left < right) {
                sum = nums[i] + nums[left] + nums[right];

                if(sum == target) {
                    return sum;
                }

                if(Math.abs(sum - target) < closest) {
                    ret = sum;
                    closest = Math.abs(sum - target);
                }

                if(sum < target) {
                    left++;
                } else if (sum > target) {
                    right--;
                }
            }
        }

        return ret; 
    }
}

File: 16-3sum-closest/3sum-closest.py
class Solution:
    def threeSumClosest(self, nums: List[int], target: int) -> int:
        nums = sorted(nums)
        ret = 0

        # same as regular 3Sum solution except every single time, just check whether it is in fact
        # the closest distance to the target by having a closestValue marker

        closestValue = float('inf')
        newSum = 0

        for i in range(len(nums) - 2):
            left = i + 1
            right = len(nums) - 1
            while left < right:
                currentSum = nums[left] + nums[right] + nums[i]
                if closestValue > abs(target - currentSum):
                    closestValue = abs(target - currentSum)
                    newSum = currentSum
                
                if target < currentSum:
                    right -= 1
                elif target > currentSum:
                    left += 1
                else:
                    left += 1
                    right -= 1
        return newSum 